third_party/move/tools/move-linter/src/model_ast_lints/known_to_abort.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements an expression linter that checks for expressions
+//! that are known to abort at compile time.
+//!
+//! The following cases are checked:
+//! * `x << n`, `x >> n` where `n` is a constant >= the bit width of `x`'s type
+//! * `x / 0`, `x % 0` (divide or modulo by zero)
+//! * `constant as type` where `constant` is outside the range of `type`
+
+use move_compiler_v2::external_checks::ExpChecker;
+use move_model::{
+    ast::{
+        Exp,
+        ExpData::{self, Call, Value},
+        Operation,
+        Value::Number,
+    },
+    model::GlobalEnv,
+    ty::Type,
+};
+use num::BigInt;
+use Operation::*;
+
+const DIVISION_BY_ZERO_MSG: &str = "Division by zero will abort";
+const MODULO_BY_ZERO_MSG: &str = "Modulo by zero will abort";
+const SHIFT_OVERFLOW_MSG: &str =
+    "Shift by amount greater than or equal to the type's bit width will abort";
+const CAST_OVERFLOW_MSG: &str =
+    "Cast operation will abort because the value is outside the target type's range";
+
+#[derive(Default)]
+pub struct KnownToAbort;
+
+fn get_constant_value(args: &Vec<Exp>, op: &Operation) -> Option<BigInt> {
```

Comment:
We might want to add some comments here saying we currently handle expressions involving `+`, `*`, and const integer values, but this could be extended to include all numerical and bitwise operations. In fact, that is when this lint is likely the most useful, as it might be obvious to see aborts with a const value, but maybe not so with a computed const expression.